managed/services/telemetry/datasource_grafana_sqlitedb.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package telemetry provides telemetry functionality.
+package telemetry
+
+import (
+	"context"
+	"database/sql"
+	"io"
+	"os"
+
+	// Driver for for grafana sqlite database.
+	_ "github.com/mattn/go-sqlite3"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type dsGrafanaSelect struct {
+	log      *logrus.Entry
+	config   DSGrafanaSqliteDB
+	db       *sql.DB
+	tempFile string
+}
+
+// check interfaces.
+var (
+	_ DataSource = (*dsGrafanaSelect)(nil)
+)
+
+// Enabled flag that determines if data source is enabled.
+func (d *dsGrafanaSelect) Enabled() bool {
+	return d.config.Enabled
+}
+
+// NewDataSourceGrafanaSqliteDB makes new data source for grafana sqlite database metrics.
+func NewDataSourceGrafanaSqliteDB(config DSGrafanaSqliteDB, l *logrus.Entry) DataSource {
+	return &dsGrafanaSelect{
+		log:      l,
+		config:   config,
+		db:       nil,
+		tempFile: "",
+	}
+}
+
+func (d *dsGrafanaSelect) Init(ctx context.Context) error {
+	// validate source file db
+	sourceFileStat, err := os.Stat(d.config.DBFile)
+	if err != nil {
+		return err
+	}
+
+	if sourceFileStat.Size() == 0 {
+		return errors.Errorf("Sourcefile %s is empty.", d.config.DBFile)
+	}
+
+	if !sourceFileStat.Mode().IsRegular() {
+		return errors.Wrapf(err, "%s is not a regular file", d.config.DBFile)
+	}
+
+	source, err := os.Open(d.config.DBFile)
+	if err != nil {
+		return err
+	}
+
+	tempFile, err := os.CreateTemp(os.TempDir(), "grafana")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := source.Close(); err != nil {
+			d.log.Errorf("Error closing file. %s", err)
+		}
+	}()
+
+	nBytes, err := io.Copy(tempFile, source)
```

Comment:
Why do we need to copy a database file?

Reply:
Because SQLite database file is can be corrupted.
The problem with file based sqlite3 database is the concurrency. The database can be easily corrupted if two processes are using the same file.  Official site says that db should be threadsafe, but many users complained about their corrupted databases.

Reply:
AFAIK SQLite handles concurrent reads gracefully and has problems with concurrent writes, leading to lock contention. Anyway, looks good.

Reply:
Yes, but we have no full control on sql driver behavior. Thank you.

Reply:
We still should consider that copy of DB may be corrupted: https://www.sqlite.org/howtocorrupt.html

Reply:
Still - better to have corrupted copy of database and having not sent telemetry than corrupt the original one :)